contrib: Finish Docker documentation.

davecgh opened the pull request:
**This is rebased on #3044**.

This updates the `contrib/docker/README.md` and `contrib/docker/Dockerfile` to address the remaining outstanding TODOs and finish fleshing out the documentation.

jholdstock commented on contrib/docker/Dockerfile:
```diff
@@ -34,19 +30,19 @@ RUN adduser \
     CGO_ENABLED=0 GOOS=linux \
     go install -trimpath -tags safe,netgo,timetzdata \
       -ldflags="-s -w" \
       . ./cmd/gencerts ./cmd/promptsecret
```
Why bother building `gencerts` and `promptsecret`?

davecgh replied:
My thought is that both of them are needed on a fairly regular basis.  `gencerts` is used if you want to use clientcert authentication (which I know a few people who do) and `promptsecret` is needed with `dcrctl` if you ever want to feed it a password.

jholdstock replied:
Ok so its "just in case". I thought I might have missed something and they are depended on by something.

jholdstock commented on contrib/docker/Dockerfile:
```diff
@@ -34,19 +30,19 @@ RUN adduser \
+    go install -trimpath -tags safe,netgo -ldflags="-s -w"
 
 # Build entrypoint helper for the production image.
 WORKDIR /go/src/github.com/decred/dcrd/contrib/docker/entrypoint
```
Why run `go mod init` here? Cant a `go.mod` and `go.sum` file for entrypoint be checked into git?

davecgh replied:
I didn't want to check them in because it would make it a separate module in the repository that all of the tools that look for submodules would pick up and confuse them.

davecgh replied:
I forgot to mention that also, I figured that if we ever want to start offering images on e.g. dockerhub via a separate repo, the code for `entrypoint` would need to live in that other repository and thus having a `go.mod` would result in the wrong path.  Creating it during the build is simpler and avoids those types of issues.

jholdstock commented on contrib/docker/README.md:
````diff
@@ -101,105 +90,322 @@ value.
-   $ docker start "${DCRD_CONTAINER_NAME}"
-   ```
+```sh
+$ docker stop "${DCRD_CONTAINER_NAME}"
````
[docker stop](https://docs.docker.com/engine/reference/commandline/stop/) will send a `SIGTERM`, and if the container hasn't stopped in 10 seconds it sends a `SIGKILL`. Might want to extend that timeout, e.g. for 60 seconds:

```sh
$ docker stop -t 60 "${DCRD_CONTAINER_NAME}"
```

jholdstock commented on contrib/docker/entrypoint/entrypoint.go:
```diff
@@ -70,8 +77,7 @@ func main() {
 		decredData := os.Getenv("DECRED_DATA")
```
The two lines to declare `decredData` and `dcrdAppData` are duplicated and can be moved up out of the switch

davecgh commented:
Thanks for the review!

> ...the concept of running as 10000:10000 is explained in detail twice,

Good point.  Updated to put a link in the overview to the section discussion this to avoid that repetition.

> the dcrctl RPC examples are listed twice...

I believe they are all different with the exception of running the container where the important aspect about the DNS names are explained in the `RPC Server Authentication` section.  As you mentioned that is so there can be a quick start section and then the reader can find out the details of why.

I suppose the `RPC Server Authentication` could link to the `Quick Start` section for the example, but I personally find that style where I have to jump around a lot annoying as a reader when I'm first reading about a new system.  I'm not fundamentally opposed to changing that if others prefer jumping around instead though.

> Defaulting to always build latest master seems a bit off to me. I believe most users would want to be running a release version and defaulting to the latest release would be more sensible. I notice you've listed that as a TODO though, so I guess that can be left for now.

Yeah, I wasn't 100% sure on that point since I don't really use docker and why I requested a review from someone who does.  I would imagine that it will highly depend on the user which is why I added the TODO since I figured having a choice would cover the various use cases.  For example, I pretty much always run `master` on all of the core Decred software, so being unable to do that would be one more reason not to use Docker.

Another thing to keep in mind about building a release is that they typically are fairly tightly coupled to the Go version that was the latest at the time of the release, so I suspect that having a release version would really end up needing to have a directory per release version that pulls the specific Go version to build with too.

jholdstock reviewed (approved):
Updates look good.

I guess which version to build and which go version to use really depends on the target audience. Devs probably want latest, people deploying for production probably want release. We can maybe address both just by using params (default to release because devs by their nature are more comfortable changing things via params). That should satisfy devs and also make deployment onto things like Umbrel/Unraid/Proxmox/etc nice and simple.
